src/plugin/ogc/wfs/wfslayerconfig.js

```diff
@@ -352,94 +357,113 @@ plugin.ogc.wfs.WFSLayerConfig.prototype.addMappings = function(layer, options) {
 
 
 /**
- * @type {Array<RegExp>}
+ * The available WFS type config objects. Plugins can add supported parser types to this.
+ * @type {Array<os.ogc.WFSTypeConfig>}
  */
-plugin.ogc.wfs.WFSLayerConfig.PREFERRED_TYPES = [
-  /^application\/json$/,
-  /gml\/?3/i,
-  /gml\/?2/i
+plugin.ogc.wfs.WFSLayerConfig.TYPE_CONFIGS = [
+  {
+    regex: /^application\/json$/,
+    parser: 'geojson',
+    type: 'geojson',
+    priority: 100
+  },
+  {
+    regex: /gml\/?3/i,
+    parser: 'gml',
+    type: 'gml3',
+    priority: 50
+  },
+  {
+    regex: /gml\/?2/i,
+    parser: 'gml',
+    type: 'gml2',
+    priority: -100
+  }
 ];
 
 
 /**
- * @enum {number}
+ * Register a type config object.
+ * @param {!os.ogc.WFSTypeConfig} config
+ * @protected
  */
-plugin.ogc.wfs.WFSLayerConfig.TYPES = {
-  GEOJSON: 0,
-  GML3: 1,
-  GML2: 2
+plugin.ogc.wfs.WFSLayerConfig.registerType = function(config) {
+  const configs = plugin.ogc.wfs.WFSLayerConfig.TYPE_CONFIGS;
+
+  if (!configs.find((c) => c.type === config.type)) {
+    configs.push(config);
+  }
+
+  // sort them by priority, highest first
+  configs.sort((a, b) => goog.array.defaultCompare(b.priority, a.priority));
 };
 
 
 /**
  * @param {Object<string, *>} options
- * @return {number} 0 = JSON, 1 = GML3, 2 = GML2
+ * @return {Object} The type config to use.
  * @protected
  */
 plugin.ogc.wfs.WFSLayerConfig.prototype.getBestType = function(options) {
   var formats = /** @type {Array<string>|undefined} */ (options['formats']);
   var format = /** @type {string} */ (this.params.get('outputformat'));
-  var preferred = plugin.ogc.wfs.WFSLayerConfig.PREFERRED_TYPES;
+  var preferred = plugin.ogc.wfs.WFSLayerConfig.TYPE_CONFIGS;
 
   // see if the given format is one mutually supported by the layer and this plugin
   if (format && (!formats || formats.includes(format))) {
     for (var i = 0, n = preferred.length; i < n; i++) {
-      var regex = preferred[i];
-      if (regex.test(format)) {
-        return i;
+      var pref = preferred[i];
+      if (pref.regex.test(format)) {
+        return pref;
       }
     }
   }
 
   // otherwise, check the available formats for one we support and set that on the params
   if (formats) {
-    for (i = 0, n = preferred.length; i < n; i++) {
-      regex = preferred[i];
+    for (var i = 0, n = preferred.length; i < n; i++) {
+      var pref = preferred[i];
 
       for (var j = 0, m = formats.length; j < m; j++) {
-        if (regex.test(formats[j])) {
+        if (pref.regex.test(formats[j])) {
           if (this.params) {
             this.params.set('outputformat', formats[j]);
           }
 
-          return i;
+          return pref;
         }
       }
     }
   }
 
   this.params.remove('outputformat');
-  return 1;
+  return preferred[1];
```

Comment:
This was intended to make GML3 the default format, which it will do unless `registerType` is called and index 1 changes. We may want to assign the predefined configs as their own objects, reference those objects in the `TYPE_CONFIGS` definition, and return the GML3 type here.

Reply:
Agreed. This was from my testing where I was pushing a higher priority one in. For the sake of my direct object comparison tests, having those as constants would be good.